File: mercury-imap/src/main/java/org/abstracthorizon/mercury/imap/util/FlagUtilities.java
/*
 * Copyright (c) 2004-2007 Creative Sphere Limited.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *
 *   Creative Sphere - initial API and implementation
 *
 */
package org.abstracthorizon.mercury.imap.util;

import javax.mail.Flags;

/**
 * Utilities for flags
 *
 * @author Daniel Sendula
 */
public class FlagUtilities {

    /**
     * Returns string representation of flags
     * @param flags flags
     * @return  // DateHolder
     */
    public static String toString(Flags flags) {
        if (flags == null) {
            return "";
        }
        StringBuffer buf = new StringBuffer();
        Flags.Flag[] fs = flags.getSystemFlags();
        boolean first = true;
        if ((fs != null) && (fs.length > 0)) {
            for (int i=0; i<fs.length; i++) {
                if (i > 0) {
                    buf.append(' ');
                }
                if (fs[i].equals(Flags.Flag.ANSWERED)) {
                    buf.append("\\Answered");
                } else if (fs[i].equals(Flags.Flag.DELETED)) {
                    buf.append("\\Deleted");
                } else if (fs[i].equals(Flags.Flag.FLAGGED)) {
                    buf.append("\\Flagged");
                } else if (fs[i].equals(Flags.Flag.SEEN)) {
                    buf.append("\\Seen");
                } else if (fs[i].equals(Flags.Flag.DRAFT)) {
                    buf.append("\\Draft");
                } else if (fs[i].equals(Flags.Flag.RECENT)) {
                    buf.append("\\Recent");
                }
            } // for
            first = false;
        }
        String[] ss = flags.getUserFlags();
        if ((ss != null) && (ss.length > 0)) {
            for (int i=0; i<ss.length; i++) {
                if (!first) {
                    buf.append(',');
                } else {
                    first = false;
                }
                buf.append(ss[i]);
            } // for
        }
        return buf.toString();
    }

    /**
     * Returns maildir representation of flags
     * @param flags flags
     * @return maildir string of flags
     */
    public static String toMaildirString(Flags flags) {
        if (flags == null) {
            return "";
        }
        StringBuffer buf = new StringBuffer();
        if (flags.contains(Flags.Flag.DRAFT)) {
            buf.append('D');
        }
        if (flags.contains(Flags.Flag.FLAGGED)) {
            buf.append('F');
        }
        if (flags.contains(Flags.Flag.ANSWERED)) {
            buf.append('R');
        }
        if (flags.contains(Flags.Flag.SEEN)) {
            buf.append('S');
        }
        if (flags.contains(Flags.Flag.DELETED)) {
            buf.append('T');
        }
        return buf.toString();
    }
}

File: mercury-maildir-ext/src/main/scala/org/abstracthorizon/mercury/maildir/scala/util/FlagUtilities.scala
package org.abstracthorizon.mercury.maildir.scala.util

import javax.mail.Flags;

object FlagUtilities {
    /**
     * Creates maildir string from given flags.
     * <ul>
     * <li>&quot;D&quot; for DRAFT</li>
     * <li>&quot;F&quot; for FLAGGED</li>
     * <li>&quot;R&quot; for ANSWERED</li>
     * <li>&quot;S&quot; for SEEN</li>
     * <li>&quot;T&quot; for DELETED</li>
     * </ul>
     *
     * @param flags flags to be converted
     * @return string in maildir format
     */
    def toMaildirString(flags: Flags): String = {
        if (flags == null) {
            return ""
        }
        val buf = new StringBuffer()
        if (flags.contains(javax.mail.Flags.Flag.DRAFT)) {
            buf.append('D')
        }
        if (flags.contains(javax.mail.Flags.Flag.FLAGGED)) {
            buf.append('F')
        }
        if (flags.contains(javax.mail.Flags.Flag.ANSWERED)) {
            buf.append('R')
        }
        if (flags.contains(javax.mail.Flags.Flag.SEEN)) {
            buf.append('S')
        }
        if (flags.contains(javax.mail.Flags.Flag.DELETED)) {
            buf.append('T')
        }
        return buf.toString()
    }

    /**
     * Creates set of flags from maildir flags string.
     * <ul>
     * <li>For &quot;D&quot; adds DRAFT</li>
     * <li>For &quot;F&quot; adds FLAGGED</li>
     * <li>For &quot;R&quot; adds ANSWERED</li>
     * <li>For &quot;S&quot; adds SEEN</li>
     * <li>For &quot;T&quot; adds DELETED</li>
     * </ul>
     * @param str maildir flags string
     * @return flags set
     */
    def fromMaildirString(str: String): Flags = {
        val flags = new Flags()
        val len = str.length()
        
        if ((str == null) || (len == 0)) {
            return flags
        }
        for (i <- 0 to (len - 1)) {
            val c = Character.toUpperCase(str.charAt(i))
            if (c == 'D') {
                flags.add(Flags.Flag.DRAFT)
            } else if (c == 'F') {
                flags.add(Flags.Flag.FLAGGED)
            } else if (c == 'R') {
                flags.add(Flags.Flag.ANSWERED)
            } else if (c == 'S') {
                flags.add(Flags.Flag.SEEN)
            } else if (c == 'T') {
                flags.add(Flags.Flag.DELETED)
            }
        }
        return flags
    }
    
    def main(args: Array[String]) = {
        println("Hello world 1")
    }

}

class FlagUtilities {

    def main(args: Array[String]) = {
        println("Hello world 2")
    }
    
}